### ISSUE TITLE
False positive DL3042

### ISSUE THREAD

**Issue by antonagestam** [hadolint/hadolint#497]
- [ ] This is a bug report
- [x] This is a feature request
- [x] I searched existing issues before opening this one

Hi! We use `--mount=type=cache,target=/root/.cache/pip` for our pip installation step, however hadolint doesn't recognize this and errors with `DL3042`, suggesting we should use `--no-cache-dir`. If hadolint tries to be smart enough to find problems inside calls, it would be nice if it also recognizes that a cache volume is mounted here.

### Expected behavior

No `DL3042` error for pip installation without `--no-cache-dir` but with `--mount=type=cache,target=/root/.cache/pip`.

### Actual behavior

`DL3042` error for pip installation without `--no-cache-dir` but with `--mount=type=cache,target=/root/.cache/pip`.

**Comment by lorenzo** [comment]
can you show the full dockerfile? or at least the relevant section we can use to reproduce the issue?

**Comment by antonagestam** [comment]
Here's the relevant section:

```
RUN --mount=type=cache,target=/root/.cache/pip \
    set -x && \
    pip install pip==$PIP_PIP_VERSION pip-tools==$PIP_PIP_TOOLS_VERSION && \
    pip install --require-hashes --pre -r reqs/requirements.txt && \
    if [ "${BUILD_TARGET}" = "dev" ] ; then \
        pip install --require-hashes --pre -r reqs/dev-requirements.txt ; \
    fi && \
    pip check
```

**Comment by 4x0v7** [comment]
This is possibly a separate issue, but the "problematic" and "correct" examples for DL3042 in the wiki appear to be backwards

**Comment by shadowspawn** [comment]
For anyone else who didn't recognise the syntax, `--mount` is an experimental syntax for BuildKit :

- https://docs.docker.com/engine/reference/builder/#buildkit
- https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/experimental.md

**Comment by mwgamble** [comment]
This `--mount` syntax is no longer experimental. It would be great to see support for this added :pray:

**Comment by aiven-anton** [comment]
@mwgamble That's pretty cool, and actually should probably change what the recommended fix for this violation is :)

**Comment by florian-sattler** [comment]
Are there any updates on this? I would like to stop ignoring `DL3042` as it is very useful in general.

**Comment by slimslenderslacks** [comment]
I've added pull request for DL3060 ( #1003 ), which is a similar issue but for `yarn install`.  As @mwgamble mentioned, the `--mount` syntax is no longer valid.  

If after review, this pattern works for DL3060, then we can apply it to DL3042 as well.

**Comment by antonagestam** [comment]
> As @mwgamble mentioned, the --mount syntax is no longer valid.

I think this was maybe a typo? What he wrote is it's no longer _experimental_.

**Comment by slimslenderslacks** [comment]
That's right.  I've updated my comment.  The `--mount` syntax is no longer _experimental_.

**Comment by KUGA2** [comment]
This is fixed for apt (DL3009) but not for pip (DL3042). Correct?
